### PR TITLE
Remove enum picongpu::CommunicationTag

### PR DESCRIPTION
**Pull request by sbastrakov**
This was noticed during work on #3558. But since that one was a bugfix and to be backported, while the removal required some refactoring as well I separated it into this PR. Said removal required modifying and modernizing several places. I didn't split this modernizing part into several commits or PRs as I didn't initially estimate how many changes were required. There are two more cleanup things I will do as follow-ups, will mark it in the code.

### CODE REVIEW

File: include/picongpu/fields/MaxwellSolver/Yee/Yee.hpp [hunk 1]
```diff
@@ -152,18 +152,28 @@ namespace picongpu
 
     namespace traits
     {
-        template<class CurlE, class CurlB>
```
Comment (sbastrakov) [inline include/picongpu/fields/MaxwellSolver/Yee/Yee.hpp:155]: Note: I think this had no reason to be in the .def file to begin with. It just happened to work before, and wouldn't now.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.def [hunk 2]
```diff
@@ -34,20 +34,4 @@ namespace picongpu
 
         } // namespace maxwellSolver
     } // namespace fields
-
-    namespace traits
-    {
-        template<typename T_CurlE, typename T_CurlB>
```
Comment (sbastrakov) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.def:40]: Note: I think this had no reason to be in the .def file to begin with. It just happened to work before, and wouldn't now.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp [hunk 3]
```diff
@@ -461,6 +461,33 @@ namespace picongpu
...
+         *
+         * @tparam T_CurlE functor to compute curl of E
+         * @tparam T_CurlB functor to compute curl of B
+         */
+        template<typename T_CurlE, typename T_CurlB>
+        struct GetMargin<picongpu::fields::maxwellSolver::YeePML<T_CurlE, T_CurlB>, FieldE>
+            : public GetMargin<picongpu::fields::maxwellSolver::Yee<T_CurlE, T_CurlB>, FieldE>
+        {
+        };
+
```
Comment (sbastrakov) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp:488]: Note: as a follow-up PR I will add `GetMargin` version for field solvers without the field to give max from `FieldE` and `FieldB` - we have a use case for that doing it manually.

File: include/picongpu/traits/GetMargin.hpp [hunk 4]
```diff
@@ -31,10 +31,10 @@ namespace picongpu
          * \tparam Solver solver which needs ghost cells for solving a problem
          *         if solver not define `LowerMargin` and `UpperMargin` this trait (GetMargin)
          *         must be specialized
-         * \tparam SubSetName a optional name (id) if solver needs different ghost cells
+         * \tparam T_Parameter an optional parameter type
          * for different objects
          */
-        template<class Solver, unsigned int SubSetName = 0>
+        template<class Solver, typename T_Parameter = void>
```
Comment (sbastrakov) [inline include/picongpu/traits/GetMargin.hpp:37]: I will modernize this file and add `T_Parameter` to `GetLowerMargin` and `GetUpperMargin` as a follow-up PR.

Comment (sbastrakov) [reply include/picongpu/traits/GetMargin.hpp:37]: Unrelated note: there were no other use cases for the integer template parameter, and anyhow type parameter is just more general, one could use an integral constrant type if needed there.

**Comment by sbastrakov** [comment]
Okay @psychocoderHPC , this is the second time our documented `git clang-format` workflow failed to format a change in a `.def` tile. I think something is wrong there.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.def (hunk 2 above)
Comment (psychocoderHPC) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.def:40]: IMO it was in the def file because of include order issues.

**Comment by psychocoderHPC** [comment]
CI is failing because one GPU is dead. I connected the admins already and will restart the failing jobs as soon as the CI is restarted.

**Comment by sbastrakov** [comment]
@psychocoderHPC now in clang one test seems to hang up: `/usr/bin/mpiexec "--allow-run-as-root" "-n" "1" "./PMaccTest-memory-3D"`, appears to be the same for all failed tests.

File: include/picongpu/traits/GetMargin.hpp (hunk 4 above)
Comment (sbastrakov) [reply include/picongpu/traits/GetMargin.hpp:37]: The modernization submitted in #3564.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp (hunk 3 above)
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp:488]: This is implemented in #3571 .

**Pull request merged**